web/src/settings_preferences.ts

```diff
@@ -12,41 +13,60 @@ import * as loading from "./loading";
 import * as overlays from "./overlays";
 import {page_params} from "./page_params";
 import * as settings_components from "./settings_components";
+import type {GenericUserSettings} from "./settings_config";
 import * as settings_ui from "./settings_ui";
 import {realm} from "./state_data";
 import * as ui_report from "./ui_report";
 import {user_settings} from "./user_settings";
+import type {UserSettings} from "./user_settings";
+
+type SettingsPanel = {
+    container: string;
+    settings_object: GenericUserSettings;
+    for_realm_settings: boolean;
+    notification_sound_elem: string | null;
+};
+
+type UserSettingsProperty = keyof UserSettings;
 
 const meta = {
     loaded: false,
 };
 
-export let user_settings_panel;
+export let user_settings_panel: SettingsPanel;
 
-export let user_default_language_name;
+export let user_default_language_name: string;
 
-export function set_default_language_name(name) {
+export function set_default_language_name(name: string): void {
     user_default_language_name = name;
 }
 
-function change_display_setting(data, $status_el, success_msg_html, sticky) {
-    const status_is_sticky = $status_el.data("is_sticky");
+function change_display_setting(
+    data: Record<string, string | number | boolean>,
+    $status_el: JQuery,
+    success_msg_html?: string,
+    sticky?: boolean,
+): void {
+    const status_is_sticky = $status_el.attr("data-is_sticky") === "true";
     const display_message_html = status_is_sticky
-        ? $status_el.data("sticky_msg_html")
+        ? $status_el.attr("data-sticky_msg_html")
         : success_msg_html;
-    const opts = {
-        success_msg_html: display_message_html,
-        sticky: status_is_sticky || sticky,
-    };
+    const opts =
```

Comment:
`success_msg_html` ( as a result `display_message_html`) and `sticky` are optional. Since the `success_msg_html` and `sticky` fields of `opts` object( ie. passed to `settings_ui.do_settings_change`)  can be optional but not undefined, I have changed the way `opts` is created. We only use these fields when are not `undefined`, if they are we don't pass them into the `opts` object. This was done to address the [exactOptionalPropertyTypes](https://www.typescriptlang.org/tsconfig/#exactOptionalPropertyTypes) rule enabled recently.

Reply:
[discussion](https://chat.zulip.org/#narrow/stream/6-frontend/topic/typescript.20migration/near/1800871)

Reply:
Your change does not preserve the old behavior, as removes `false` and `""`, not just `undefined`. In general, you should never rely on truthiness and falsiness tests unless the only values are `true` and `false`.

With almost all exactOptionalPropertyTypes errors, it will be simpler to change the type rather than the code.

```diff
 type RequestOpts = {
-    success_msg_html?: string;
+    success_msg_html?: string | undefined;
     failure_msg_html?: string;
     success_continuation?: (response_data: unknown) => void;
     error_continuation?: (xhr: JQuery.jqXHR) => void;
-    sticky?: boolean;
+    sticky?: boolean | undefined;
     $error_msg_element?: JQuery;
 };
```

Reply:
Okay -- made the suggested changes.

Reply:
Got it, thanks! -- made the suggested changes.